### DISCUSSION TITLE
FS25 Progress, status, infos

### DISCUSSION THREAD

**Discussion by Tensuko** [Courseplay/Courseplay_FS22#3600] [Announcements]
Here we will share information about our progress on migrating Courseplay to FS25.

While @pvaiko edited the code, so that we can even get into the game, @schwiti6190 have started with getting the UI up and running again.

![grafik](https://github.com/user-attachments/assets/817a0be7-0a76-4006-acad-2e26c3136068)
![grafik](https://github.com/user-attachments/assets/9355f693-1ebe-42c9-89e9-bf1a3ba5b07d)

We can not give you any date when we get Courseplay running again, but I will keep you informed about our progress.

Courseplay is a free, community project so donations are never expected but always appreciated. If you wish to support our work with more than kind words, check our [Help us out page](https://github.com/Courseplay/Courseplay_FS22?tab=readme-ov-file#help-us-out). You can even buy us a coffee now :)

**Comment by bruceyboy24804** [comment]
Its looking good guys, keep up the great work. I've always appreciated the amount of work and effort and time you guts put in to make a great mod like this work.

**Comment by bojanh66** [comment]
Thank you for your hard work, this is the mod that i cant play without.
Did you ever get an offer from Giants to work with them for all these years that CP exist ?
I mean, talented team like you guys, that can make way better AI than Giants and you are still independent team is really suprise, so many talented modders that could improve vanilla FS, CP, Autodrive, Autoload and other script modders.

**Reply by Tensuko** [reply]
Well, we didn't get any offer and since we have all our normal daily work to do, working with them will be complicated.
However, I offered them to show and explain them what CP can do and how it works - they had no interest.
We keep push CP forward and show them how it could work :)

**Reply by fudoido** [reply]
They had no interest, and since fs22 they are trying to make something similar in the base game.....
As for working "with them" would be complicated, i think the best option would be you guys being full time employees with a decent wage, that should be the honest proposal from giants.
I have seen a few euro truck modders joining SCS as devs for either scripting experience or moddeling experience, wouldnt be surprising if giants decided to take the same route, as i agree with Bojanh66 and you deserve a place in the official dev team of FS.
Honest opinion apart now and talking about the mod, hope the code doesnt need too much changes apart from the gui elements and we can get the full courseplay options as soon it gets released for fs25, and by options i mean the fieldwork, shovel, silo work, and the main features we got so much used to.
Keep going guys, you are the most popular mod team ever in the fs community.

**Reply by Tensuko** [reply]
We all have our jobs, there is no way we will work for Giants.

**Comment by Tensuko** [comment]
Making progress... (still a lot to do...)
![grafik](https://github.com/user-attachments/assets/736e10a6-4583-4087-9f01-c934c6bc1e04)

**Reply by bruceyboy24804** [reply]
is the conversion from 19 to 22 or 22 to 25 more difficult... because surely it cant be as bad a from 19 to 22?

**Reply by Tensuko** [reply]
Hard to say yet, 19 to 22 was a reconstruction and went pretty good and was necessary.
I guess thats why we don't have to hard time to convert it to 25, but still thats just the beginning.

**Reply by bruceyboy24804** [reply]
i still miss 19s ui so convenient :(

**Reply by bruceyboy24804** [reply]
im also curious what your plans for hold h stuff is

**Reply by Tensuko** [reply]
FS19 UI was crap and unmaintainable.
Hold H button nothing is planed.
We may use the hold "any key" button.
I have something in mind for that, but let us finish get stuff working again first then we can see what we can adjust and improve :)

**Reply by BoxerHUN** [reply]
Dear developers!
I already know how to use it and I love AutoDrive. But... if only because of the precise parking, but I think I can say on behalf of many that we would very much welcome the possibility of creating individual routes. With all due respect, please consider this option as well.

https://www.youtube.com/watch?v=v3xQfoWy9sM

**Reply by pvaiko** [reply]
> is the conversion from 19 to 22 or 22 to 25 more difficult... because surely it cant be as bad a from 19 to 22?

Does not look bad. The course generator works as it has no dependency on the Giants code. Surprisingly, it looks like they kept the AI Job/Task/Drive Strategy architecture more or less intact, which means we can also drive the generated courses, no big hiccups there so far. 

The problem is the UI, our current one is integrated into their UI which also means that we have a lot of dependencies. And they have changed a lot of other things too (not just UI, everywhere) that may need some mostly trivial changes. 

Trivial, if we had the documentation and the scripts, but they don't seem to give a damn, obviously all those script developers downstream aren't important enough to them to release the sources. 

They don't even bother to tell _when_ they will be condescend to publish those, so all we have left is reverse engineering their crap.

**Reply by bruceyboy24804** [reply]
So basically your saying if you documentation the progress would be sped up a lot faster 🤔 it's almost like the farm sim community keeps their game  a float but giants couldn't care less about helping their game because their not givings community the tools they need 🙄

**Reply by BoxerHUN** [reply]
> So basically your saying if you documentation the progress would be sped up a lot faster 🤔 it's almost like the farm sim community keeps their game a float but giants couldn't care less about helping their game because their not givings community the tools they need 🙄

Or they just sit on their laurels... pfff

**Reply by Tensuko** [reply]
> Dear developers! I already know how to use it and I love AutoDrive. But... if only because of the precise parking, but I think I can say on behalf of many that we would very much welcome the possibility of creating individual routes. With all due respect, please consider this option as well.
> 
> https://www.youtube.com/watch?v=v3xQfoWy9sM

I cant say much about that, just that we had ideas and brainstorming - but the old system wont come back. However I dont say that we have no plans... As said to other requests... Let us finish the convert first, then we see what we can do...

**Reply by bruceyboy24804** [reply]
Tensuko and the course play team S tier code programmers 😁

**Comment by coatsy35** [comment]
I know you said you don't want to do anything with pressing and holding H, but it would be cool if one of them modes was CP and the generator settings were there so it looked like part of the game. Obviously you have to wait and see what parts you can alter when they release the SDK. I just think it would look good and tie in well. Obviously you need the mini HUD as before. Be really nice if your course generator can utilise some of the Giants ability for faster pattern generation as it is fast and alters very quickly on custom fields from what I've tested, (I have merged all kinds of fields and had not issue with generation so far, simplistic as it may be).

Keep up the great work all of you. It is appreciated

**Reply by Tensuko** [reply]
> I know you said you don't want to do anything with pressing and holding H, but it would be cool if one of them modes was CP and the generator settings were there so it looked like part of the game.

Thats not a reason to stuff all our settings in a way too small window.
Imagine that map preview on a bigger field or with narrow work width, you won't see anything from the preview.
Again, will not do that. Let us do what we have planed, it will be cool, trust me.

The reason for faster generation can be:
- Generator is not in LUA
- The new LUA API Giants use

The first we won't have access to and would require redoing the generator, the 2nd is something that could already happening without we doing anything.

**Reply by bruceyboy24804** [reply]
No convenience 😅🤣

**Reply by pvaiko** [reply]
Just to clarify: there is no new Lua API, they seem to be using now Luau which is an extension of Lua by Roblox. 

They forgot to tell anyone about this too as far as I can tell, ignoring the community once again. Luau is much better than Lua as it has type checking and other features that would have saved us hundreds of hours of troubleshooting.

Also, if they really switched to Luau, that must have been planned a long time ago, and if we knew about it, we could have refactored the course generator using Luau, which, for instance, has vectors built in the language. With Lua, we had to roll our own vector class and the complete vector algebra. Lot of time wasted.

We don't know however what Giants is actually implementing from the language in their runtime. For instance, it looks like their implementation does not have coroutines, which we used to run the generator in the "background".

**Reply by bruceyboy24804** [reply]
damn, why is giants so out of touch with their community

**Reply by k4talin** [reply]
Because it seems they are a bounch of Arrogants.

**Reply by Tensuko** [reply]
Hmm dont think so....
Rather a bunch of chaotic ppl.
I mean it seems that their community coordinator has left now too but no informations about that as well....
I mean he is just there for community stuff, why communicate that with the community....

Wonder how long Benni stays now :)

**Comment by BoxerHUN** [comment]
It's finally out... https://gdn.giants-software.com/downloads.php

**Reply by bruceyboy24804** [reply]
> It's finally out... https://gdn.giants-software.com/downloads.php

still no Scripting API Reference for fs25 though

**Comment by Misamangame** [comment]
Bought FS25 just to have that 'bonus' thing but still playing FS22. Without CP & AD this game is simply useless so I hope I did not waste my money but I'm risk taker :). Keep up the good work and take as much time as you need I'm patitent :)

**Comment by pvaiko** [comment]
A quick status, but first of all we'd like to thank you for your generous donations, we really appreciate it. 🙏🏻 

Giants published the Studio 10.0.0 today, so we do now have a debugger, but the scripts and documentation are still not available, so that still prevents us from properly implementing stuff, such as automatic working width detection.

* Course generator is working. We we'll look into later if we can leverage the new Giants code, for instance to detect field and island boundaries. 
* Course generator UI partly works, no preview yet and have some trouble connecting the buttons.
* Direct start works
* Driving fieldwork courses, turns, implement on/off also seems to work fine.
* Bale collection works.
* Saving/loading vehicle courses works.

One major issue is that for whatever reason Giants removed the coroutine package from the Lua engine. We relied on that to keep the pathfinder working in the "background" (in reality, it wasn't running in parallel, it is just preemptive multitasking) to prevent the game from freezing while the pathfinder is running. We'll either find a replacement in their engine or have to implement something for it, right now, the pathfinder freezes the game.

**Reply by Tensuko** [reply]
![grafik](https://github.com/user-attachments/assets/6f35f47c-689c-4018-97ee-1998616b15ea)

**Reply by bruceyboy24804** [reply]
does that cause freezing

**Reply by pvaiko** [reply]
What I meant by "freeze" is that the game stops while the pathfinder is running (if it is running long enough for you to notice), but then it continues normally.

**Reply by bruceyboy24804** [reply]
ah i thought you meant not responding

**Reply by blubstar86** [reply]
Does `Bale collection works.` also means bale wrapping works?
Ty for the close look of the progress you make.

**Reply by Tensuko** [reply]
I think so. Will try it out as well, but there are still bugs to solve anyway.

**Reply by bruceyboy24804** [reply]
> I think so. Will try it out as well, but there are still bugs to solve anyway.

Thank you and the rest of the group that helped make this great mod come to life

**Reply by antonkapralov** [reply]
> A quick status, but first of all we'd like to thank you for your generous donations, we really appreciate it. 🙏🏻
> 
> Giants published the Studio 10.0.0 today, so we do now have a debugger, but the scripts and documentation are still not available, so that still prevents us from properly implementing stuff, such as automatic working width detection.
> 
> * Course generator is working. We we'll look into later if we can leverage the new Giants code, for instance to detect field and island boundaries.
> * Course generator UI partly works, no preview yet and have some trouble connecting the buttons.
> * Direct start works
> * Driving fieldwork courses, turns, implement on/off also seems to work fine.
> * Bale collection works.
> * Saving/loading vehicle courses works.
> 
> One major issue is that for whatever reason Giants removed the coroutine package from the Lua engine. We relied on that to keep the pathfinder working in the "background" (in reality, it wasn't running in parallel, it is just preemptive multitasking) to prevent the game from freezing while the pathfinder is running. We'll either find a replacement in their engine or have to implement something for it, right now, the pathfinder freezes the game.

is where a chance you will release some beta version, so we can / could help you with bug / performance / functionality testing ?
we can / could send you logfiles / bug reports .

**Reply by Tensuko** [reply]
Only when we make the first release.
Getting tons of wild feedback wont help us now in this state.

**Reply by antonkapralov** [reply]
> Only when we make the first release. Getting tons of wild feedback wont help us now in this state.

Understandable ! Just wanted to help, i started to play FS22 and bought FS25 only becouse of you guys and in hope to be able to play FS25 with CoursePlay and Autodrive. Just cant wait till you make it work. i am sorry have some lack of patience with waiting. i a msorry ones again. you really made FS22 much more injoyble to play it BIG. Honestly Giants should take you and your team in the company and add courseplay / autodrive as part of the game itself (maybe as extra DLC). dont undestand why they cant / dont wont to make in game AI similaar to your script. 

Sorry again.

**Reply by pvaiko** [reply]
> Honestly Giants should take you and your team in the company

Oh no way, I have a daytime job I really enjoy, building systems on a scale much larger than a game...

**Reply by pvaiko** [reply]
> One major issue is that for whatever reason Giants removed the coroutine package from the Lua engine. We relied on that to keep the pathfinder working in the "background" (in reality, it wasn't running in parallel, it is just preemptive multitasking) to prevent the game from freezing while the pathfinder is running.

Ok, this is resolved, I implemented a workaround, pathfinder now works as in FS22, you'll see an FPS drop, but no freeze.

**Reply by bruceyboy24804** [reply]
Damn, I'm fluent in c# but not lua never even touched lua

**Reply by MoslessNeo** [reply]
The CP has improved a lot since the team "remodeled" and "organized" the code. Mainly for operation with auger wagon (overloader) when the tractor needs to find a route to the harvester or truck. The FPS drops are due to the calculation that is made but this is not a problem.

**Reply by antonkapralov** [reply]
> > Honestly Giants should take you and your team in the company
> 
> Oh no way, I have a daytime job I really enjoy, building systems on a scale much larger than a game...

Cool you still work after that on project like this. Funny to see how ppl outside of the company making game better then the makers of the game itself does, and sadly to see them dont wonting to make their own game better by implementing soch a great code likes yours in the original game by themself. Tnank you pvaiko for doing such a great job.

**Comment by StefanBraun2001** [comment]
Oh, so there will be a Courseplay for this Farming Sim. Will be very helpful for vegetable harvesting with that 3.0 metre self propelled harvester because its a pain with the integrated AI helper. And with bringing stuff from A to B without the AI helper stating the target is "unreachable". Will we also be able to drive the combine harvester (of any kind) and call a trailer to unload is (e. g. by honking for 3 seconds)? Because I'd rather like to drive the combine harvester than the trailer...

**Reply by k4talin** [reply]
Haven't hear anything about AutoDrive? Do we know anything about it?

**Reply by Tensuko** [reply]
In their DC they made an announcement.

**Reply by k4talin** [reply]
who the heck got their DC. Zero info on github.

**Reply by k4talin** [reply]
> In their DC they made an announcement.

Can u help me w alink or info please?

**Reply by Tensuko** [reply]
AD github and scroll down to the info box.

**Reply by k4talin** [reply]
> AD github and scroll down to the info box.

ty  vm

**Comment by bojanh66** [comment]
Guys, how is the CP fps performance in FS25 when you hire 3-4 workers, does fps dips ? Is FS25 better in CPU utilization when running CP because of newer game engine ?
In FS22 it was quite bad when i was running 4-5 workers, my fps would go from 100 to maybe 40-50 and i have 11600k OC and rtx 3070ti OC.

**Reply by Tensuko** [reply]
Dont know yet, got to get the things working first before we consider such testing.

**Reply by pvaiko** [reply]
Is 40 FPS in farming simulator really a problem?

**Reply by coatsy35** [reply]
If it’s over 30 I’m usually quite happy 😆 it’s not an FPS

**Reply by antonkapralov** [reply]
> If it’s over 30 I’m usually quite happy 😆 it’s not an FPS

honestly i think it is not the GPU issuer, i guess this is more CPU problem, as it needs to do lots of calculation in background. Maybe some memory leaks.

**Reply by Misamangame** [reply]
> Guys, how is the CP fps performance in FS25 when you hire 3-4 workers, does fps dips ? Is FS25 better in CPU utilization when running CP because of newer game engine ? In FS22 it was quite bad when i was running 4-5 workers, my fps would go from 100 to maybe 40-50 and i have 11600k OC and rtx 3070ti OC.

you must have absolutely badly optimized system or played shitty map with drop like that with only 4-5 workers.  I was running 40 workers like 10-15CP and rest AD and had significant drops only when more vehicles started pathfinding.  And I had AMD FX6300.  Generally I watched about 1 FPS drop per 3-4 workers. And yes when they started pathfinding I was on 20-25 temporarily , but not with normal work. Why you expecting FS 25 will be better? Name 1 game (FS style or city builder style) optimized by developers - 0.  It will be always CPU heavy

**Reply by antonkapralov** [reply]
> > Guys, how is the CP fps performance in FS25 when you hire 3-4 workers, does fps dips ? Is FS25 better in CPU utilization when running CP because of newer game engine ? In FS22 it was quite bad when i was running 4-5 workers, my fps would go from 100 to maybe 40-50 and i have 11600k OC and rtx 3070ti OC.
> 
> you must have absolutely badly optimized system or played shitty map with drop like that with only 4-5 workers. I was running 40 workers like 10-15CP and rest AD and had significant drops only when more vehicles started pathfinding. And I had AMD FX6300. Generally I watched about 1 FPS drop per 3-4 workers. And yes when they started pathfinding I was on 20-25 temporarily , but not with normal work. Why you expecting FS 25 will be better? Name 1 game (FS style or city builder style) optimized by developers - 0. It will be always CPU heavy

Agree with you, optimization is not best friends of modern Developers right now. Look at FS25 right now, loots of texture / lighting issure at this moment, and dont beter qua optimization as FS22 .

**Reply by k4talin** [reply]
> Is 40 FPS in farming simulator really a problem?

Hell NOOO. It is not Formula 1 nor Counter Strike.

**Comment by pvaiko** [comment]
Quick update: as already mentioned, running the pathfinder in the background is working now, as do most vehicles. 

We have issues with some of the new ones, mainly the harvesters, won't properly start. It is difficult to tell why only those, and reverse engineering is very time consuming and isn't really fun. Only if we had those scripts ...

**Reply by k4talin** [reply]
Thx for updates.

**Reply by Roby1164** [reply]
I noticed that the helper of the game, it often happened to me when the grass needs to be mowed, if the field is not completely and perfectly sown, it does not start to do the job, I don't know if it can be helpful information .

**Reply by Tensuko** [reply]
That is becouse Giants detects the position where it should work based on the ground and fruit.
Thats not a problem for us, we can even mow sand :)

**Reply by k4talin** [reply]
> That is becouse Giants detects the position where it should work based on the ground and fruit. Thats not a problem for us, we can even mow sand :)

I liked how they avoid to learn from you when you did offer to teach them. And here we go. They created a abomination.

**Reply by pvaiko** [reply]
Ok, vehicle startup bug is fixed now, there may a few outliers where we can't accurately measure the pipe length, but the worst is over...

Also, they published some of the scripts with the latest patch, although I wasn't able to load them in the debugger.

**Comment by Roby1164** [comment]
Am I wrong or with this FS25 update, they have inserted the indications for the scripts, collisions and all the rest??

**Reply by Tensuko** [reply]
Ya we know and it is also written in their changelog.

**Comment by k4talin** [comment]
@pvaiko Any updates SIR? Are you getting close? LOVE

**Reply by Tensuko** [reply]
Some bugs and callstacks got fixed.
Continue working on the UI, not yet rdy to show you, but we are getting there.
Still some stuff to solve and to find the right way.

Getting closer on each step :)

Update: Something to look at.
![grafik](https://github.com/user-attachments/assets/2243fc33-e222-4f7a-b29b-fc36534cb5a4)

**Reply by k4talin** [reply]
Super. Thank you

**Comment by Tensuko** [comment]
To not keep you in the dark for too long.... My plan is to make an uncommented video about what we have so far, even if I get into bugs. Hope I get it done this weekend.

**Reply by Misamangame** [reply]
Great! Still honest follower of your YT unless there is some new one :)

**Comment by Tensuko** [comment]
Uncut Testing Session.
https://youtu.be/M1DfqweMhg8

**Reply by BigBoyDe** [reply]
Respect.

**Reply by Podvorie** [reply]
Great!

**Reply by k4talin** [reply]
TY

**Reply by CaptainRush** [reply]
nice. i also supported via paypal.

If you need beta testers im willing to test

**Reply by Tensuko** [reply]
Thank you very much.
We will have a public beta when we have obvious bugs fixed and a "known issues" list.

**Reply by Isikana** [reply]
Looks great. The harvester driving at around 13 min makes me sad for crop destruction tho :D

The rest was faboulous :)

**Comment by Tensuko** [comment]
Target found :)
![grafik](https://github.com/user-attachments/assets/1c15f0b2-07ed-4ee6-b7fa-60673b3c3d8e)

**Reply by MEminK66** [reply]
That looks great. Can't wait for public beta.

**Comment by fudoido** [comment]
From the video and the images posted, seems you guys are advancing well in the new version. Hope all major bugs get sorted soon and giants releases some more info for you guys meanwhile.

**Comment by pvaiko** [comment]
https://github.com/Courseplay/Courseplay_FS25

**Reply by Ursus95** [reply]
You beauty.
